Add PipeSource and PipeTarget implementation for native piping directly to/from parent process

DaRosenberg opened the pull request:
This PR dresses issue #79 by providing an explicit mode whereby streams can be directly piped (using native `Process` functionality) to/from the parent process.

Instead of opening the parent process's streams and creating `StreamPipeSource/StreamPipeTarget`, calling code can use `PipeSource.ParentProcess` and `PipeTarget.ParentProcess` respectively, to indicate that the affected child process stream should be natively piped through to its corresponding parent process counterpart.

Parent process can now do this:

```cs
var cmd =
  Cli.Wrap("git")
    .WithWorkingDirectory(@"/Users/danielrosenberg/Git/StacksTest")
    .WithArguments("mergetool")
    .WithValidation(CommandResultValidation.None)
    .WithStandardInputPipe(PipeSource.ParentProcess)
    .WithStandardOutputPipe(PipeTarget.ParentProcess)
    .WithStandardErrorPipe(PipeTarget.ParentProcess);

var result = await cmd.ExecuteAsync();

var key = Console.ReadKey(intercept: true); // This now works as expected
Console.WriteLine($"Read key: {key.KeyChar}");
```

Just a draft for now to show the basic idea, please let me know what needs to be added to make it an acceptable contribution. I'm guessing test coverage might be one thing.

Fixes #79

Tyrrrz commented:
Thank you for the effort.

It appears some tests have failed, but looking at the implementation I'm still not convinced it's worth it regardless. This is adding quite a lot of complexity and branching logic for one specific use case. Also, the problem appears only with stdin, so these workaround are not needed for stdout/stderr.

I think a better solution would be to make a PR into dotnet/runtime that adds cancellation support for `Console.GetStandardInput()`. Alternatively, you can make a fork of this library that allows you to modify the generated `ProcessStartInfo` just in time before the process is ran. I'm not planning to expose `Process` internals in the library itself as it will break encapsulation.

DaRosenberg commented:
Maybe there's still a chance I might convince you, so I'll try. ;)

While the change is indeed not strictly needed for STDOUT/STDERR, I added it for 2 reasons:
- Doing it this way (presumably) has fewer moving parts and better perf characteristics, since no work needs to be performed by the application to copy between streams
- Consistency/discoverability (developer does not need to apply completely different approaches for input vs output)

I look at this as less of a workaround, and more like the most efficient approach for a separate case (the 5 cases now being in-memory, stream, null, command and parent process). It makes sense to me that the latter constitutes a separate case.

But also, do you really think this is such an edge case? For a console app to want to run a child process in "interactive shell" mode seems like quite a core scenario... you don't agree?

The test that broke is probably because I removed the workaround you had put in place from before, but actually maybe it's best that it stays in because there may be other non-cancellable stream sources where the side-effects of the lingering copy task is less detrimental.

Tyrrrz commented:
Sorry for late response, I was giving it some more thought.

> But also, do you really think this is such an edge case? For a console app to want to run a child process in "interactive shell" mode seems like quite a core scenario... you don't agree?

It was actually one of those scenarios that I wasn't considering at all. I would be surprised to find out that it's common. In most cases, I imagine, you would be running CLI in a deterministic way, where the required inputs (args, env vars, stdin) are known and prepared beforehand.

There are a few things that bother me with this change:

1. The new `PipeSource`/`PipeTarget` is essentially a "marker" object and the handling logic actually resides in `Command`. This approach conflicts with the design that allows pipes to be interchangeable and composable by having their behavior encapsulated in their own objects. The `PipeTarget.ParentProcess` also won't work with `PipeTarget.Merge` because of that, meaning that only the base `ExecuteAsync` will work and not any of the alternate execution models.

2. The behavior where the streams are routed to the parent process is contingent on whether the spawned subprocess inherits the same stream handles as the parent. This is true by default in .NET as of right now, but [it could change in the future](https://github.com/dotnet/runtime/issues/13943). Additionally, this would only work if the host process is a console application. If it's not, then the spawned process will have its own console window and, by not redirecting the pipes, all data will be written to it, which is much slower than piping to `null`. So in the end, `PipeTarget.ParentProcess` will not even work correctly, as that will pipe data to an entirely different process. Also, since this behavior is not documented, it's not a good idea to rely on it, in my opinion.

Again, you're welcome to use this fork for your needs, but I believe that this change is too impactful to be merged into the library.

DaRosenberg commented:
You make some very good points there.

In fact, the inability to compose using `PipeTarget.Merge` also became clear to me since my last comment, because we actually have a need to do that in our program.

We will use our fork for the time being, and if I think of some clever way to implement this change without the shortcomings you've highlighted, I'll get in touch.

Thanks for the careful consideration! 👍

Tyrrrz commented:
Sounds good!
